geopandas/io/parquet.py

```diff
@@ -0,0 +1,362 @@
+
+    if not isinstance(df, DataFrame):
+        raise ValueError("to_parquet only supports IO with DataFrames")
+
+    # must have value column names (strings only)
+    if df.columns.inferred_type not in {"string", "unicode", "empty"}:
+        raise ValueError("parquet must have string column names")
+
+    # index level names must be strings
+    valid_names = all(
+        isinstance(name, str) for name in df.index.names if name is not None
+    )
+    if not valid_names:
+        raise ValueError("Index level names must be strings")
+
+
+def to_parquet(df, path, compression="snappy", index=None, **kwargs):
+    """
+    Write a GeoDataFrame to the parquet format.
+
+    Any geometry columns present are serialized to WKB format in the file.
+
+    Requires 'pyarrow'.
+
+    Parameters
+    ----------
+    path : str
+        File path or Root Directory path. Will be used as Root Directory path
+        while writing a partitioned dataset.
+    compression : {'snappy', 'gzip', 'brotli', None}, default 'snappy'
+        Name of the compression to use. Use ``None`` for no compression.
+    index : bool, default None
+        If ``True``, include the dataframe's index(es) in the file output. If
+        ``False``, they will not be written to the file. If ``None``, the
+        engine's default behavior will be used.
+    kwargs
+        Additional keyword arguments passed to the engine
+    """
+
+    import_optional_dependency(
+        "pyarrow.parquet", extra="pyarrow is required for parquet support."
+    )
+    from pyarrow import parquet, Table, Schema
+
+    _validate_dataframe(df)
+
+    geometry_columns = df.columns[df.dtypes == "geometry"].tolist()
+    geometry_column = df._geometry_column_name
+
+    # Encode CRS info for serialization
+    # TODO: getattr() is a shim until GeoSeries have crs attributes (#1193);
+    # defaults to CRS of GeoDataFrame if missing.
+    crs = {
+        col: _encode_crs(getattr(df[col], "crs", df.crs)) for col in geometry_columns
+    }
+
+    # Convert to a DataFrame so we can convert geometries to WKB while
+    # retaining original column names.
+    df = DataFrame(df.copy())
+
+    path, _, _, _ = get_filepath_or_buffer(path, mode="wb")
+
+    # Encode all geometry columns to WKB
+    for col in geometry_columns:
+        df[col] = to_wkb(df[col].values)
+
+    if index is None:
+        from_pandas_kwargs = {}
+    else:
+        from_pandas_kwargs = {"preserve_index": index}
+
+    schema = Schema.from_pandas(df, **from_pandas_kwargs)
+
+    # Store geopandas specific column level metadata
+    # This must be done BEFORE creating the parquet Table to be able to
+    # persist the values
+    for col in geometry_columns:
+        # only write CRS metadata if defined
+        if crs[col] is None:
+            continue
+
+        field = schema.field(col).with_metadata({"crs": _encode_metadata(crs[col])})
+        index = schema.get_field_index(col)
+        schema = schema.set(index, field)
```

Comment:
OK, I see now what you meant in https://github.com/geopandas/geopandas/pull/1191#discussion_r353333972 (I should have looked at your updated implementation before!)

So you are storing the column level metadata in the arrow field (column) metadata in the *arrow* schema. 
The problem with this, though, is that this *arrow* field metadata is not translated to the *parquet* column metadata (see https://issues.apache.org/jira/browse/ARROW-4359).

The latest version of pyarrow actually stores the full arrow schema in the parquet file metadata (in a "ARROW:schema" field, and pyarrow this way restores the schema when reading, but this is only since pyarrow 0.15), so the information is still available in the parquet file, but it is somewhat "hidden" somewhere inside the serialized arrow schema (which makes it not very accessible to potentially other applications or packages).

So given the above, I think I might prefer storing the column information together in the "geo" file metadata (similar to how pandas does it: https://dev.pandas.io/docs/development/developer.html)